### PR TITLE
Chainer: New Package.

### PR DESCRIPTION
**Pull request by tkameyama**

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-chainer/package.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+    It provides automatic differentiation APIs based on the define-by-run
+    approach (a.k.a. dynamic computational graphs) as well as object-oriented
+    high-level APIs to build and train neural networks.
+    It also supports CUDA/cuDNN using CuPy for high performance training
+    and inference.
+    """
+
+    homepage = "https://chainer.org/"
+    url      = "https://github.com/chainer/chainer/archive/v7.2.0.tar.gz"
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-chainer/package.py:21]: You can add me as a maintainer for this package:
```python
maintainers = ['adamjstewart']
```
I maintain all of the other DL packages in Spack, so why not 1 more?

File: var/spack/repos/builtin/packages/py-chainer/package.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+    version('7.2.0', sha256='6e2fba648cc5b8a5421e494385b76fe5ec154f1028a1c5908557f5d16c04f0b3')
+    version('6.7.0', sha256='87cb3378a35e7c5c695028ec91d58dc062356bc91412384ea939d71374610389')
+
+    depends_on('python@3.5:', when='@7:', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-chainer/package.py]:
```suggestion
    depends_on('python@3.5.1:', when='@7:', type=('build', 'run'))
```
They have a bunch of logic to fail if the version is exactly 3.5.0, might as well respect it.

File: var/spack/repos/builtin/packages/py-chainer/package.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    depends_on('py-setuptools', type='build')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-chainer/package.py]:
```suggestion
    depends_on('py-setuptools', type=('build', 'run'))
```

File: var/spack/repos/builtin/packages/py-chainer/package.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    depends_on('py-numpy@1.9:', type='run')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-chainer/package.py]:
```suggestion
    depends_on('py-numpy@1.9:', type=('build', 'run'))
```

File: var/spack/repos/builtin/packages/py-chainer/package.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    depends_on('py-six', type='run')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-chainer/package.py]:
```suggestion
    depends_on('py-six@1.9.0:', type=('build', 'run'))
```

File: var/spack/repos/builtin/packages/py-chainer/package.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    depends_on('py-typing-extensions', type='run')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-chainer/package.py]:
```suggestion
    depends_on('py-typing-extensions', type=('build', 'run'))
```

File: var/spack/repos/builtin/packages/py-chainer/package.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    depends_on('py-filelock', type='run')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-chainer/package.py]:
```suggestion
    depends_on('py-filelock', type=('build', 'run'))
```

File: var/spack/repos/builtin/packages/py-chainer/package.py [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    # for pkg_resouces
+    depends_on('py-pip', type='run')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-chainer/package.py]: `pkg_resources` comes from `setuptools`, not `pip`. You can remove this dependency

File: var/spack/repos/builtin/packages/py-chainer/package.py [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    version('7.2.0', sha256='6e2fba648cc5b8a5421e494385b76fe5ec154f1028a1c5908557f5d16c04f0b3')
+    version('6.7.0', sha256='87cb3378a35e7c5c695028ec91d58dc062356bc91412384ea939d71374610389')
+
+    depends_on('python@3.5:', when='@7:', type=('build', 'run'))
+    depends_on('py-setuptools', type='build')
+    depends_on('py-numpy@1.9:', type='run')
+    depends_on('py-six', type='run')
+    depends_on('py-typing-extensions', type='run')
+    depends_on('py-filelock', type='run')
+    # for pkg_resouces
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-chainer/package.py]: Missing some dependencies:
```python
depends_on('py-protobuf@3:', type=('build', 'run'))
depends_on('py-typing@:3.6.6', when='@:6', type=('build', 'run'))
```

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-chainer/package.py (hunk 9 above)
Comment (tkameyama) [reply var/spack/repos/builtin/packages/py-chainer/package.py]: Chainer seems to  depend on py-typing-extention@:3.6.6 in setup.py.
But py-typing-extention@3.6.6 is not found in package.py.
Also py-typing-extentions was depends py-typing@3.7.4:
So I change py-typing-extensions.

File: var/spack/repos/builtin/packages/py-typing-extensions/package.py [hunk 10]
```diff
@@ -17,7 +17,14 @@ class PyTypingExtensions(PythonPackage):
...
+    version('3.6.6', sha256='51e7b7f3dcabf9ad22eed61490f3b8d23d9922af400fe6656cb08e66656b701f')
 
     depends_on('python@2.7:2.8,3.4:', type=('build', 'run'))
     depends_on('py-setuptools', type='build')
-    depends_on('py-typing@3.7.4:', when='^python@:3.4', type=('build', 'run'))
+    depends_on(
+        'py-typing@3.7.4:', when='@3.7: ^python@:3.4', type=('build', 'run')
+    )
+    depends_on(
+        'py-typing@3.6.0:3.6.9',
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-typing-extensions/package.py]:
```suggestion
        'py-typing@3.6.2:',
```

File: var/spack/repos/builtin/packages/py-typing-extensions/package.py [hunk 11]
```diff
@@ -17,7 +17,14 @@ class PyTypingExtensions(PythonPackage):
...
+        when='@3,6 ^python@:3.4', type=('build', 'run')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-typing-extensions/package.py]:
```suggestion
        when='^python@:3.4', type=('build', 'run')
```

File: var/spack/repos/builtin/packages/py-chainer/package.py [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+    depends_on('py-numpy@1.9:', type=('build', 'run'))
+    depends_on('py-six@1.9.0:', type=('build', 'run'))
+    depends_on('py-typing-extensions', type=('build', 'run'))
+    depends_on(
+        'py-typing-extensions@:3.6.6', when='@:6', type=('build', 'run')
+    )
+    depends_on('py-filelock', type=('build', 'run'))
+    depends_on('py-protobuf@3:', type=('build', 'run'))
+    depends_on(
+        'py-typing@:3.6.6', when='@:6 ^python@:3.6', type=('build', 'run')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-chainer/package.py]:
```suggestion
        'py-typing@:3.6.6', when='@:6', type=('build', 'run')
```
The dependency in the package is unconditional, even if it is built into Python at that version.

File: var/spack/repos/builtin/packages/py-chainer/package.py [hunk 13]
```diff
@@ -0,0 +1,42 @@
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+# ----------------------------------------------------------------------------
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-chainer/package.py]: You can delete this line

**Review by adamjstewart — CHANGES_REQUESTED** [review]
I would put the `depends_on` statements on a single line. We ignore `depends_on` lines in our flake8 line length checks.

File: var/spack/repos/builtin/packages/py-chainer/package.py (hunk 12 above)
Comment (tkameyama) [reply var/spack/repos/builtin/packages/py-chainer/package.py]: But I install py-chainer@6.7.0 on python 3.7.7, `import chainer `is failed as following:
<pre>
$ python
Python 3.7.7 (default, Apr 23 2020, 09:47:45) 
[GCC 4.8.5 20150623 (Red Hat 4.8.5-39)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import chainer
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/kameyama/work/test/spack/opt/spack/linux-centos7-sandybridge/gcc-4.8.5/py-chainer-6.7.0-qzxc647vt63afqnckoyzlrc3e6cpt7v3/lib/python3.7/site-packages/chainer/__init__.py", line 10, in <module>
    from chainer import backends  # NOQA
  File "/home/kameyama/work/test/spack/opt/spack/linux-centos7-sandybridge/gcc-4.8.5/py-chainer-6.7.0-qzxc647vt63afqnckoyzlrc3e6cpt7v3/lib/python3.7/site-packages/chainer/backends/__init__.py", line 1, in <module>
    from chainer.backends import cuda  # NOQA
  File "/home/kameyama/work/test/spack/opt/spack/linux-centos7-sandybridge/gcc-4.8.5/py-chainer-6.7.0-qzxc647vt63afqnckoyzlrc3e6cpt7v3/lib/python3.7/site-packages/chainer/backends/cuda.py", line 37, in <module>
    import typing as tp  # NOQA
  File "/home/kameyama/work/test/spack/opt/spack/linux-centos7-sandybridge/gcc-4.8.5/py-typing-3.6.6-bnizzxvhfaapqifgtgj7aooalbjtenn7/lib/python3.7/site-packages/typing.py", line 1356, in <module>
    class Callable(extra=collections_abc.Callable, metaclass=CallableMeta):
  File "/home/kameyama/work/test/spack/opt/spack/linux-centos7-sandybridge/gcc-4.8.5/py-typing-3.6.6-bnizzxvhfaapqifgtgj7aooalbjtenn7/lib/python3.7/site-packages/typing.py", line 1004, in __new__
    self._abc_registry = extra._abc_registry
AttributeError: type object 'Callable' has no attribute '_abc_registry'
>>> 
</pre>
I add depends_on('^python@:3.6') for version 6.

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-chainer/package.py]: Hmm, in that case I guess we can keep the dependency how you had it. Whatever works. I don't know how many people will even be using the older version of Chainer anyway.

Comment (tkameyama) [reply var/spack/repos/builtin/packages/py-chainer/package.py]: I add old version to use on python2.

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**